server/proxy/server.js

```diff
@@ -0,0 +1,61 @@
+const client = require('socket.io-client')
+const express = require('express')
+const bodyParser = require('body-parser');
+const cookieParser = require('cookie-parser')
+const cors = require('cors');
+const axios = require('axios')
+const app = express()
+
+let socket
+let cards;
+
+app.use(bodyParser.json());
+app.use(cors());
+app.use(cookieParser());
+
+app.get('/connect-and-play', (req, res) => {
+    const url = req.query.url;
+
+    socket = client(url);
+
+    const instance = axios.create({
+        baseURL: url,
+        timeout: 500,
+    });
+    
+    socket.on('connect', () => {
+        instance.post('auth/login', { 
+            username: "halfling" 
+        })
+        .then(() => {
+            return instance.get('api/cards')
+        })
+        .then((response) => {
+            cards = response.data;
```

Comment:
assigning `cards` to a variable in a parent closure widens its lifecycle.

if all you wanted to do was pass it to the next promise in the chain, there are better patterns for this

Reply:
here's one (of many) ways:
```js
.then((response) => Promise.all([
	Promise.resolve(response.data),
	instance.get('api/start')
	]))
.then(([cards, startResponse]) => {
    if (startResponse.status === 200) {
        return instance.post('api/playCardWord', {
            card: cards[0].id.toString(),
            word: "Hello"
        })
    }
}) 
```

Reply:
Yeah, that's all I was using it for. Thanks, that feels a lot cleaner!

Reply:
or if a Promise pyramid is acceptable to you, you can use the Promise's closure:
```js
.then((response) => {
    cards = response.data;
    return instance.get('api/start')
    .then((startResponse) => {
	    if (startResponse.status === 200) {
	        return instance.post('api/playCardWord', {
	            card: cards[0].id.toString(),
	            word: "Hello"
	        })
	    }
	}) 
})
```
but this makes it a bit more complicated to reason about `catch()` behaviour